File: engines/calibration/height_calibration.py
"""engines/calibration/height_calibration.py — backend height-based
scale calibration.

Mirrors motionlens-web/lib/calibration/heightCalibration.ts so live
mode and upload mode produce the same pixels_per_cm for the same
clip + patient height.

Algorithm:
  1. Sample N frames from the early "standing window" of the clip
     (first ~3 s). The patient is expected to stand straight, fully
     in frame, before the test movement begins.
  2. For each sampled frame: compute body pixel height = lowest
     visible foot landmark y − highest visible head landmark y.
  3. Drop frames where torso landmarks aren't visible or where the
     measurement is anomalous compared to the rest.
  4. Take the MEDIAN of the surviving readings as the locked body
     pixel height — robust to single-frame jitter.
  5. pixels_per_cm = body_pixel_height / (patient_height_cm × 0.87)
     The 0.87 ratio is the anthropometric average for nose-to-floor
     fraction of total stature; we use it as the bridge between the
     pixel span we can measure and the patient's true height.
"""
from __future__ import annotations

import logging
from statistics import median as _stdlib_median
from typing import Any, Optional

import cv2

log = logging.getLogger(__name__)


NOSE_TO_FLOOR_HEIGHT_FRACTION = 0.87  # mirrors lib/calibration/types.ts
MIN_HEIGHT_CM = 80.0
MAX_HEIGHT_CM = 230.0

# Visibility gate for the head + foot landmark search.
_HEAD_FOOT_VIS_THRESHOLD = 0.30

# Sampling window — patient is assumed to be standing straight in
# the first few seconds of the clip.
_STANDING_WINDOW_SEC = 3.0
_DEFAULT_SAMPLE_FRAMES = 12

# Acceptable spread between agreeing samples (fraction of median).
_STABLE_TOLERANCE_FRACTION = 0.05

_HEAD_LANDMARKS = (
    "nose",
    "left_eye_inner", "left_eye", "left_eye_outer",
    "right_eye_inner", "right_eye", "right_eye_outer",
    "left_ear", "right_ear",
)

_FOOT_LANDMARKS = (
    "left_ankle", "right_ankle",
    "left_heel", "right_heel",
    "left_foot_index", "right_foot_index",
)

_TORSO_REQUIRED = (
    "left_shoulder", "right_shoulder",
    "left_hip", "right_hip",
)


def _measure_body_pixel_height(ts: dict, frame_idx: int) -> Optional[float]:
    """Body pixel height at frame_idx, computed from the smoothed
    time-series produced by engines.gait_engine.build_time_series.

    Returns None when torso isn't fully visible OR neither head nor
    feet landmarks are above the visibility threshold.
    """
    # Torso anchor must be present — otherwise the patient is
    # probably partial in frame and the height reading would be
    # meaningless.
    for name in _TORSO_REQUIRED:
        track = ts.get(name)
        if not track:
            return None
        if frame_idx >= len(track["vis"]):
            return None
        if float(track["vis"][frame_idx]) < _HEAD_FOOT_VIS_THRESHOLD:
            return None

    top_y: Optional[float] = None
    for name in _HEAD_LANDMARKS:
        track = ts.get(name)
        if not track or frame_idx >= len(track["vis"]):
            continue
        if float(track["vis"][frame_idx]) < _HEAD_FOOT_VIS_THRESHOLD:
            continue
        y = float(track["y_px"][frame_idx])
        if top_y is None or y < top_y:
            top_y = y
    if top_y is None:
        return None

    bottom_y: Optional[float] = None
    for name in _FOOT_LANDMARKS:
        track = ts.get(name)
        if not track or frame_idx >= len(track["vis"]):
            continue
        if float(track["vis"][frame_idx]) < _HEAD_FOOT_VIS_THRESHOLD:
            continue
        y = float(track["y_px"][frame_idx])
        if bottom_y is None or y > bottom_y:
            bottom_y = y
    if bottom_y is None:
        return None

    body_px = bottom_y - top_y
    return body_px if body_px > 0 else None


def measure_body_pixel_height_from_time_series(
    ts: dict,
    fps: float,
    total_frames: int,
    standing_window_sec: float = _STANDING_WINDOW_SEC,
    sample_frames: int = _DEFAULT_SAMPLE_FRAMES,
) -> Optional[float]:
    """Sample the standing-window frames and return a stable median
    body pixel height, or None when no agreeing reading was found.
    """
    if fps <= 0 or total_frames <= 0:
        return None
    window_frames = min(total_frames, int(fps * standing_window_sec))
    if window_frames < sample_frames:
        sample_frames = max(1, window_frames)
    step = max(1, window_frames // max(1, sample_frames))
    indices = list(range(0, window_frames, step))[:sample_frames]
    if not indices:
        return None

    readings: list[float] = []
    for i in indices:
        v = _measure_body_pixel_height(ts, i)
        if v is not None and v > 0:
            readings.append(v)

    if not readings:
        return None

    median = float(_stdlib_median(readings))
    # Keep only the readings within ±tolerance of the median — drops
    # outliers from partial-body frames the visibility gate let
    # through.
    inliers = [
        r for r in readings
        if abs(r - median) / median <= _STABLE_TOLERANCE_FRACTION
    ]
    if not inliers:
        # No agreement — refuse to calibrate; safer than picking a
        # noisy value.
        log.info(
            "height_calibration: readings ranged from %.0f to %.0f px "
            "(median %.0f) — no agreement, returning None.",
            min(readings), max(readings), median,
        )
        return None
    return float(_stdlib_median(inliers))


def pixels_per_cm_from_height(
    body_pixel_height_px: float,
    patient_height_cm: float,
) -> Optional[float]:
    if body_pixel_height_px <= 0:
        return None
    if patient_height_cm < MIN_HEIGHT_CM or patient_height_cm > MAX_HEIGHT_CM:
        return None
    effective_cm = patient_height_cm * NOSE_TO_FLOOR_HEIGHT_FRACTION
    if effective_cm <= 0:
        return None
    return float(body_pixel_height_px / effective_cm)


def build_height_calibration_dict(
    body_pixel_height_px: float,
    patient_height_cm: float,
    source_frame: dict[str, int],
) -> Optional[dict[str, Any]]:
    """Assemble a CalibrationResult-shaped dict for the API layer."""
    ppc = pixels_per_cm_from_height(body_pixel_height_px, patient_height_cm)
    if ppc is None:
        return None
    return {
        "pixels_per_cm": ppc,
        "detected_at_ms": 0,
        "source": "height_based",
        "patient_height_cm": float(patient_height_cm),
        "body_pixel_height_px": float(body_pixel_height_px),
        "source_frame_px": source_frame,
    }


def probe_source_frame_dimensions(video_path: str) -> dict[str, int]:
    """Best-effort read of the video's frame dimensions for the
    CalibrationResult.source_frame_px field. Returns {0,0} on failure.
    """
    cap = cv2.VideoCapture(video_path)
    try:
        w = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH) or 0)
        h = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT) or 0)
    finally:
        cap.release()
    return {"width": w, "height": h}

File: motionlens-web/lib/calibration/heightCalibration.ts
// Height-based scale calibration.
//
// How it works:
//   1. Doctor enters / confirms the patient's standing height in cm.
//   2. Patient stands straight, fully in frame.
//   3. The system samples the patient's body PIXEL height — from the
//      highest visible head landmark (nose / eye / ear) down to the
//      lowest visible foot landmark (heel / ankle / foot_index) — and
//      requires the reading to be stable for a short averaged window
//      before locking in.
//   4. pixels_per_cm = body_pixel_height / (height_cm × NOSE_TO_FLOOR_HEIGHT_FRACTION)
//      The anthropometric correction accounts for the fact that the
//      nose (our highest reliable head landmark) sits at ~87 % of
//      total stature for adults.
//
// This module is consumer-agnostic — Functional Reach uses it today,
// and D3 Single-Leg Hop / D4 CMJ will reuse it without duplication.

import type { Keypoint } from "@tensorflow-models/pose-detection";
import { LM_LIVE as LM } from "@/lib/pose/landmarks-live";
import {
  NOSE_TO_FLOOR_HEIGHT_FRACTION,
  type CalibrationResult,
} from "./types";

/** Score below which a head/foot keypoint is treated as occluded
 *  and excluded from the body-pixel-height measurement. */
export const HEAD_FOOT_VIS_THRESHOLD = 0.35;

/** Minimum height entry accepted. Anything below this is almost
 *  certainly a typo. */
export const MIN_HEIGHT_CM = 80;
export const MAX_HEIGHT_CM = 230;

/** How long the body-pixel-height reading must stay stable before
 *  the calibration locks in (sample frames at ~6 Hz detection). */
export const STABLE_FRAMES_REQUIRED = 6;

/** Two body-pixel-height readings agree when they differ by less
 *  than this fraction. 3 % is comfortably tighter than typical
 *  MediaPipe head/foot landmark jitter, so a stable reading really
 *  is stable. */
export const STABLE_TOLERANCE_FRACTION = 0.03;

const HEAD_LANDMARK_INDICES: number[] = [
  LM.NOSE,
  LM.LEFT_EYE_INNER, LM.LEFT_EYE, LM.LEFT_EYE_OUTER,
  LM.RIGHT_EYE_INNER, LM.RIGHT_EYE, LM.RIGHT_EYE_OUTER,
  LM.LEFT_EAR, LM.RIGHT_EAR,
];

const FOOT_LANDMARK_INDICES: number[] = [
  LM.LEFT_ANKLE, LM.RIGHT_ANKLE,
  LM.LEFT_HEEL, LM.RIGHT_HEEL,
  LM.LEFT_FOOT_INDEX, LM.RIGHT_FOOT_INDEX,
];

const TORSO_REQUIRED: number[] = [
  LM.LEFT_SHOULDER, LM.RIGHT_SHOULDER,
  LM.LEFT_HIP, LM.RIGHT_HIP,
];

export interface BodyHeightReading {
  /** Body pixel height (max foot y − min head y) in source image px. */
  body_pixel_height_px: number;
  /** Highest visible head landmark y (top of body in image y-down). */
  top_y_px: number;
  /** Lowest visible foot landmark y (bottom of body in image y-down). */
  bottom_y_px: number;
}

/** Compute the patient's body pixel height from a single pose. Returns
 *  null when not enough head + foot + torso landmarks are visible. */
export function computeBodyPixelHeight(kp: Keypoint[]): BodyHeightReading | null {
  if (!kp || kp.length === 0) return null;

  // Torso anchor must be visible — otherwise we're probably looking at
  // only the upper or lower body and the height reading would be
  // meaningless.
  for (const i of TORSO_REQUIRED) {
    const p = kp[i];
    if (!p || (p.score ?? 0) < HEAD_FOOT_VIS_THRESHOLD) return null;
  }

  let topY: number | null = null;
  for (const i of HEAD_LANDMARK_INDICES) {
    const p = kp[i];
    if (!p || (p.score ?? 0) < HEAD_FOOT_VIS_THRESHOLD) continue;
    if (topY === null || p.y < topY) topY = p.y;
  }
  if (topY === null) return null;

  let bottomY: number | null = null;
  for (const i of FOOT_LANDMARK_INDICES) {
    const p = kp[i];
    if (!p || (p.score ?? 0) < HEAD_FOOT_VIS_THRESHOLD) continue;
    if (bottomY === null || p.y > bottomY) bottomY = p.y;
  }
  if (bottomY === null) return null;

  const bodyPx = bottomY - topY;
  if (bodyPx <= 0) return null;
  return {
    body_pixel_height_px: bodyPx,
    top_y_px: topY,
    bottom_y_px: bottomY,
  };
}

/** Is the patient fully in frame for calibration?  Required so the
 *  body-pixel-height reading isn't artificially truncated. Returns a
 *  human-readable "reason" hint when the answer is no, for the
 *  coaching layer. */
export function checkBodyInFrame(
  kp: Keypoint[],
  reading: BodyHeightReading | null,
  frameHeightPx: number,
): { ok: boolean; reason: string } {
  if (!reading) {
    // Discriminate between "no head" / "no feet" / "no torso" so the
    // operator gets actionable coaching instead of a generic message.
    const headSeen = HEAD_LANDMARK_INDICES.some(
      (i) => (kp[i]?.score ?? 0) >= HEAD_FOOT_VIS_THRESHOLD,
    );
    const feetSeen = FOOT_LANDMARK_INDICES.some(
      (i) => (kp[i]?.score ?? 0) >= HEAD_FOOT_VIS_THRESHOLD,
    );
    const torsoSeen = TORSO_REQUIRED.every(
      (i) => (kp[i]?.score ?? 0) >= HEAD_FOOT_VIS_THRESHOLD,
    );
    if (!torsoSeen) return { ok: false, reason: "torso_missing" };
    if (!headSeen) return { ok: false, reason: "head_missing" };
    if (!feetSeen) return { ok: false, reason: "feet_missing" };
    return { ok: false, reason: "body_partial" };
  }

  // Margin: avoid locking in when head or feet are right at the
  // frame edge (they'd be cropped and pixel-height under-reads).
  const edgeMarginPx = Math.max(8, frameHeightPx * 0.02);
  if (reading.top_y_px < edgeMarginPx) {
    return { ok: false, reason: "head_at_frame_edge" };
  }
  if (reading.bottom_y_px > frameHeightPx - edgeMarginPx) {
    return { ok: false, reason: "feet_at_frame_edge" };
  }
  return { ok: true, reason: "" };
}

/** Two body-height readings are stable iff they differ by less than
 *  STABLE_TOLERANCE_FRACTION. */
export function areReadingsStable(
  a: BodyHeightReading,
  b: BodyHeightReading,
  tolerance = STABLE_TOLERANCE_FRACTION,
): boolean {
  const ref = Math.min(a.body_pixel_height_px, b.body_pixel_height_px);
  if (ref <= 0) return false;
  return (
    Math.abs(a.body_pixel_height_px - b.body_pixel_height_px) / ref <= tolerance
  );
}

/** Pure conversion. Computes pixels_per_cm from a confirmed body
 *  pixel height (e.g. the median of the stable window) + the
 *  doctor-entered patient height in cm.
 *
 *  The anthropometric NOSE_TO_FLOOR fraction (~0.87) converts the
 *  *measured* body span (highest head landmark down to the floor)
 *  into the *true* head-top-to-floor span — i.e. the patient's
 *  standing height. */
export function pixelsPerCmFromHeight(
  bodyPixelHeightPx: number,
  patientHeightCm: number,
): number | null {
  if (
    !Number.isFinite(bodyPixelHeightPx) ||
    !Number.isFinite(patientHeightCm) ||
    bodyPixelHeightPx <= 0 ||
    patientHeightCm < MIN_HEIGHT_CM ||
    patientHeightCm > MAX_HEIGHT_CM
  ) {
    return null;
  }
  const effectiveHeightCm = patientHeightCm * NOSE_TO_FLOOR_HEIGHT_FRACTION;
  if (effectiveHeightCm <= 0) return null;
  return bodyPixelHeightPx / effectiveHeightCm;
}

/** Assemble a CalibrationResult from the locked reading + patient
 *  height. */
export function buildHeightCalibration(
  bodyPixelHeightPx: number,
  patientHeightCm: number,
  sourceFrame: { width: number; height: number },
): CalibrationResult | null {
  const ppc = pixelsPerCmFromHeight(bodyPixelHeightPx, patientHeightCm);
  if (ppc === null) return null;
  return {
    pixels_per_cm: ppc,
    detected_at_ms: Date.now(),
    source: "height_based",
    patient_height_cm: patientHeightCm,
    body_pixel_height_px: bodyPixelHeightPx,
    source_frame_px: { width: sourceFrame.width, height: sourceFrame.height },
  };
}
